Unix/http/httpclientauth.c

```diff
@@ -0,0 +1,991 @@
+        return NULL;
+    }
+
+    encode_context.size = 0;
+    encode_context.pdata = NULL;
+
+    // Call once to estimate the size.
+    int rslt = Base64Enc(token->value, token->length, EncodeSizeCallback,
+                         &encode_context);
+
+    // Then we allocate the data  and add the extra
+
+    encode_context.pdata = PAL_Malloc(encode_context.size + prefix_len + 1);
+    encode_context.size = prefix_len;
+
+    memcpy(encode_context.pdata, prefix, prefix_len);
+
+    // Call again to copy the data
+    if (rslt >= 0) {
+        rslt = Base64Enc(token->value, token->length, EncodePlaceCallback, &encode_context);
+    }
+
+    encode_context.pdata[encode_context.size] = '\0';
+
+    if (pResultLen) {
+        *pResultLen = encode_context.size;
+    }
+    return encode_context.pdata;
+}
+
+static void _getStatusMsg(OM_uint32 status_code, int status_type, gss_buffer_t statusString)
+{
+    OM_uint32 message_context;
+    OM_uint32 min_status;
+
+    message_context = 0;
+
+    do {
+        gss_display_status(&min_status,
+                           status_code, status_type, GSS_C_NO_OID, &message_context, statusString);
+    } while (message_context != 0);
+}
+
+static void _ReportError(HttpClient_SR_SocketData * self, const char *msg,
+                         OM_uint32 major_status, OM_uint32 minor_status)
+{
+    HttpClient *client = (HttpClient *) self->base.data;
+    OM_uint32 min_stat = 0;
+    HttpClientCallbackOnStatus2 callback = (HttpClientCallbackOnStatus2)(client->callbackOnStatus);
+
+    gss_buffer_desc major_err = { 0 };
+    gss_buffer_desc minor_err = { 0 };
+
+    _getStatusMsg(major_status, GSS_C_GSS_CODE, &major_err);
+    _getStatusMsg(minor_status, GSS_C_MECH_CODE, &minor_err);
+    trace_HTTP_ClientAuthFailed(major_err.value, minor_err.value);
+#if defined(CONFIG_ENABLE_WCHAR)
+    (void)Swprintf(g_ErrBuff, sizeof(g_ErrBuff), L"%s %s %s\n", msg,
+                   (char *)major_err.value, (char *)minor_err.value);
+#else
+    (void)Snprintf(g_ErrBuff, sizeof(g_ErrBuff), "%s %s %s\n", msg,
+                   (char *)major_err.value, (char *)minor_err.value);
+#endif
+
+    (*callback) (client, client->callbackData, MI_RESULT_OK, g_ErrBuff);
+    gss_release_buffer(&min_stat, &major_err);
+    gss_release_buffer(&min_stat, &minor_err);
+}
+
+static int _Base64DecCallback(const void *data, size_t length, void *callbackData)
+{
+    gss_buffer_t decoded_data = (gss_buffer_t) callbackData;
+    const unsigned char *new_decoded_data = (const unsigned char *)data;
+
+    memcpy((char *)decoded_data->value + decoded_data->length, new_decoded_data, length);
+    decoded_data->length += length;
+
+    return 0;
+}
+
+/*
+**==============================================================================
+*/
+
+/*
+      Converts the SPNEGO authorization header string to an opaque gss token
+
+*/
+
+static int _getInputToken(const char *authorization, gss_buffer_t token)
```

Comment:
Not that I'm aware of. The data structures are different, and different parts are linked.